Book/php7/extensions_design/hooks.rst

```diff
@@ -0,0 +1,285 @@
+        original_handler_var_dump(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    }
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        zend_function *original;
+
+        original = zend_hash_str_find_ptr(EG(function_table), "var_dump", sizeof("var_dump")-1);
+
+        if (original != NULL) {
+            original_handler_var_dump = original->internal_function.handler;
+            original->internal_function.handler = my_overwrite_var_dump;
+        }
+    }
+
+When overwriting a class method, the function table can be found on the
+``zend_class_entry``.::
+
+    zend_class_entry *ce = zend_hash_str_find_ptr(CG(class_table), "PDO", sizeof("PDO")-1);
+    if (ce != NULL) {
+        original = zend_hash_str_find_ptr(&ce->function_table, "exec", sizeof("exec")-1);
+
+        if (original != NULL) {
+            original_handler_pdo_exec = original->internal_function.handler;
+            original->internal_function.handler = my_overwrite_pdo_exec;
+        }
+    }
+
+Modifying the Abstract Syntax Tree (AST)
+****************************************
+
+When PHP 7 compiles PHP code it converts it into an abstract syntax tree (AST)
+before finally generating Opcodes that are persisted in Opcache. The
+``zend_ast_process hook`` is called for every compiled script"
+
+This is one of the most complicated hooks to use, because it requires perfect
+understanding of the AST possibilities. Creating an invalid AST here can cause
+weird behavior or crashes.
+
+It is best to look at example extensions that use this hook:
+
+- `Google Stackdriver PHP Debugger Extension
+  <https://github.com/GoogleCloudPlatform/stackdriver-debugger-php-extension/blob/master/stackdriver_debugger_ast.c>`_
+- Based on Stackdriver this `Proof of Concept Tracer with AST <https://github.com/beberlei/php-ast-tracer-poc/blob/master/astracer.c>`_
+
+Hooking into Script/File Compilation
+************************************
+
+Whenever a user script calls ``include``/``require`` or their counterparts
+``include_once``/``require_once`` PHP core calls the function at the pointer
+``zend_compile_file`` to handle this request. The argument is a file handle
+and the result is a ``zend_op_array``.::
+
+    zend_op_array * my_extension_compile_file(zend_file_handle *file_handle, int type);
+
+There are two extensions in PHP core that implement this hook: dtrace and
+opcache.
+
+- If you start the PHP script with the environment variable ``USE_ZEND_DTRACE``
+  and compiled PHP with dtrace support, then ``dtrace_compile_file`` is used
+  from ``Zend/zend_dtrace.c``.
+
+- Opcache stores op arrays in shared memory for better performance, so that
+  whenever a script is compiled its final op array is served from a cache and
+  not re-compiled. You can find this implementation in
+  ``ext/opcache/ZendAccelerator.c``.
+
+- The default implementation is called ``compile_file`` is part of the
+  generated scanner code in ``Zend/zend_language_scanner.c``.
```

Comment:
I think it would be more useful to refer to `zend_language_scanner.l` here, rather than the generated code.